packages/sitecore-jss-react/src/components/PlaceholderCommon.tsx

```diff
@@ -193,6 +203,19 @@ export class PlaceholderCommon<T extends PlaceholderProps> extends React.Compone
           : `component-${index}`;
         const commonProps = { key };
 
+        // if editMode is equal to 'metadata' then emit shallow chromes for hydration in Pages
+        if (this.props.sitecoreContext?.editMode === EditMode.Metadata) {
+          return (
+            <PlaceholderMetadata
```

Comment:
From my understanding the PlaceholderMetadata
should accept a minimum amount of props and we shouldn't duplicate most part of functionality from _PlaceholderCommon_
My expectation that you can render it here: https://github.com/Sitecore/jss/blob/dev/packages/sitecore-jss-react/src/components/PlaceholderCommon.tsx#L254
If your rendering has placeholders, it will render another _Placeholder_ component, so all this functionality will be executed again. But in your case you are trying to traverse all the tree at once (seems like)
Here is an example of rendering that contains a nested placeholder: https://github.com/Sitecore/jss/blob/dev/packages/create-sitecore-jss/src/templates/nextjs-styleguide/src/components/graphql/GraphQL-Layout.tsx#L31

```ts
const rendered = React.createElement<{ [attr: string]: unknown }>(
  component as React.ComponentType,
  this.props.modifyComponentProps ? 
  this.props.modifyComponentProps(finalProps) : finalProps
);

// if editMode is equal to 'metadata' then emit shallow chromes for hydration in Pages
if (this.props.sitecoreContext?.editMode === EditMode.Metadata) {
  return <PlaceholderMetadata {...someRequiredProps}>{rendered}</PlaceholderMetadata>;
}

return rendered;
```

Reply:
All other comments depend on the decision here.
Suggested approach is great but we wont be able to handle nested placeholders.
Yes I am traversing the whole tree at once in order to handle nested placeholders, please let me know if you can think of something better than this?

Reply:
@sc-addypathania Please, update your unit test as it suggested here: https://github.com/Sitecore/jss/pull/1776#discussion_r1579082115
You will see, that traversing the whole tree is a problem, since you can have some nested _Placeholders_, after you render renderings using _componentFactory_